src/frequenz/sdk/microgrid/_old_component_data.py

```diff
@@ -0,0 +1,1287 @@
+    """A private base class for strongly typed component data classes."""
+
+    component_id: ComponentId
+    """The ID identifying this component in the microgrid."""
+
+    timestamp: datetime
+    """The timestamp of when the data was measured."""
+
+    states: Set[ComponentStateCode | int] = frozenset()
+    """The states of the component."""
+
+    warnings: Set[ComponentErrorCode | int] = frozenset()
+    """The warnings of the component."""
+
+    errors: Set[ComponentErrorCode | int] = frozenset()
+    """The errors of the component."""
+
+    CATEGORY: ClassVar[ComponentCategory] = ComponentCategory.UNSPECIFIED
+    """The category of this component."""
+
+    METRICS: ClassVar[frozenset[Metric]] = frozenset()
+    """The metrics of this component."""
+
+    @abstractmethod
+    def to_samples(self: Self) -> ComponentDataSamples:
+        """Convert the component data to a component data object."""
+
+    @staticmethod
+    def _from_samples(class_: type[T], /, samples: ComponentDataSamples) -> T:
+        """Create a new instance from a component data object."""
+        if not samples.metric_samples:
+            raise ValueError("No metrics in the samples.")
+
+        # FIXME: This might not be true forever, but the service for now seems to send
+        # all metrics with the same timestamp for now.
+        timestamp = samples.metric_samples[-1].sampled_at
+        for sample in samples.metric_samples[:-1]:
+            if sample.sampled_at != timestamp:
+                _logger.warning(
+                    "ComponentData has multiple timestamps. Using the last one. Samples: %r",
+                    samples,
+                )
+                break
+
+        if not samples.states:
+            return class_(component_id=samples.component_id, timestamp=timestamp)
+
+        # FIXME: Maybe we can have more than one, in which case we need to merge them?
+        if len(samples.states) > 1:
+            _logger.warning(
+                "ComponentData has more than one state. Using the last one. States: %r",
+                samples.states,
+            )
+
+        return class_(
+            component_id=samples.component_id,
+            timestamp=timestamp,
+            states=samples.states[-1].states,
+            warnings=samples.states[-1].warnings,
+            errors=samples.states[-1].errors,
+        )
+
+    @classmethod
+    @abstractmethod
+    def from_samples(cls, samples: ComponentDataSamples) -> Self:
+        """Create a new instance from a component data object."""
+
+    @classmethod
+    def _check_category(cls, component_id: ComponentId) -> None:
+        """Check if the given component_id is of the expected type.
+
+        Args:
+            component_id: Component id to check.
+
+        Raises:
+            ValueError: if the given id is unknown or has a different type.
+        """
+        # pylint: disable-next=import-outside-toplevel,cyclic-import
+        from frequenz.sdk import microgrid
```

Comment:
why not relative path instead: `from .. import microgrid`

Reply:
I guess no reason.
```suggestion
        from .. import microgrid
```